src/target/lpc55xx.c

```diff
@@ -37,12 +37,515 @@
 #include "target_internal.h"
 #include "cortexm.h"
 
-#define LPC55xx_CHIPID 0x50000ff8U
-#define LPC55_DMAP_IDR 0x002a0000U
-
+#define LPC55_DMAP_IDR                 0x002a0000U
 #define LPC55_DMAP_BULK_ERASE          0x02U
 #define LPC55_DMAP_START_DEBUG_SESSION 0x07U
 
+#define LPC55xx_FLASH_BASE     0x00000000U
+#define LPC55xx_ERASE_KEY      0x6b65666cU
+#define LPC55xx_CHIPID_ADDRESS 0x50000ff8U
+
+#define LPC55xx_WRITE_SIZE 4096
+#define LPC55xx_ERASE_SIZE 4096
+
+/* Target memory layout for IAP calls, we will scribble over SRAM */
+
+#define LPC55xx_FLASH_CONFIG_ADDRESS 0x04000000U
+#define LPC55xx_CODE_PATCH_ADDRESS   0x0400003cU
+#define LPC55xx_UUID_ADDRESS         0x04000040U
+#define LPC55xx_WRITE_BUFFER_ADDRESS 0x20000000U
+#define LPC55xx_SCRATCH_MEMORY_LEN   0x40
+#define LPC55xx_UUID_LEN             0x10
+
+/* The ROM code seems to default to an MSP in the SRAM-X region, which
+ * is code-only. This seems fairly safe, so do the same for IAP calls.
+ *
+ * This may not matter as supposedly IAP doesn't use the stack (?)
+ */
+
+#define LPC55xx_IAP_MSP_ADDRESS 0x04003000U
+#define LPC55xx_IAP_FREQ_IN_MHZ 96
+
+/* Device chip IDs */
+
+#define LPC5502_CHIPID  0xa1003702U // UM11424
+#define LPC5504_CHIPID  0xa1001504U // UM11424
+#define LPC5506_CHIPID  0xa1000406U // UM11424
+#define LPC5512_CHIPID              // (unknown)
+#define LPC5514_CHIPID              // (unknown)
+#define LPC5516_CHIPID              // (unknown)
+#define LPC5524_CHIPID              // (unknown)
+#define LPC5526_CHIPID  0xa010221aU // NXP forum
+#define LPC5528_CHIPID  0xa010111cU // NXP forum
+#define LPC55S04_CHIPID 0x51001584U // UM11424
+#define LPC55S06_CHIPID 0x51000486U // UM11424
+#define LPC55S14_CHIPID             // (unknown)
+#define LPC55S16_CHIPID             // (unknown)
+#define LPC55S26_CHIPID 0xa010229aU // NXP forum
+#define LPC55S28_CHIPID 0xa010119cU // NXP forum
+#define LPC55S66_CHIPID             // (unknown)
+#define LPC55S69_CHIPID             // (unknown)
+
+/* The available IAP commands that we support, mostly flash access */
+
+enum iap_cmd {
```

Comment:
this could be typedef'd to `iap_cmd_e` as in
```c
typedef enum iap_cmd {
    [...]
} iap_cmd_e;
```
which would then simplify uses below (same for the iap_status enum)